### PR TITLE
Improve documentation on saga behaviour when using Sql Server transport and persistence

### PR DESCRIPTION
**Pull request by jpalac**

### CODE REVIEW

**Comment by ramonsmits** [comment]
I think we should create a separate page that sits under both SQL Server Transport and SQL Persistence and contains all information about combining these:

Menu:

```
Transports
--SQL Server
  -- SQL Persistence integration
Persistence
--SQL
  -- SQL Server transport integration
```

or maybe shorter in the menu like:

```
Transports
--SQL Server
  -- Persistence integration
Persistence
--SQL
  -- Transport integration
```

Maybe "Sharing transport and persistence" or "Combining "

I think it likely helps to put all of this in a table. I don't think this information should be put in the SQL Server Transport transaction guidance.

SQL Transport<br/>TransactionMode | SQL Persistence | Matching dialect | Outbox | Outcome
:-:|:-:|:-:|:-:|:-:
TransactionScope |  ✅| ✅|✅| Connection sharing via SQLT storage context
AtomicSendsWithReceive |  ✅| ✅|✅| SQLP uses isolated transaction
ReceiveOnly |  ✅| ✅|✅| Connection sharing via SQLT storage context
None |  ✅| ✅|✅| Connection sharing via SQLT storage context

**Comment by jpalac** [comment]
@ramonsmits I've made some adjustments - see what you think. To make it less complex I've split it into two sections: with and without outbox, used the sql dialect by default so no need for it in the table, and added a column for where the saga is actually stored so that it's spelled out for the user.

**Review by johnsimons — COMMENTED** [review]
Is this ready for a review yet?

**Comment by jpalac** [comment]
Go for it

________________________________

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+## Connection behavior
+
+When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is used or not. This influences where the saga data is stored.
+
+### Without Outbox
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| Connection sharing via SQLT storage context | Transport Db
```
Comment (abparticular) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:18]: Note, I'm not sure if &lt;sup&gt; will work on docs site

```suggestion
TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
```

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+AtomicSendsWithReceive |  ✅| SQLP uses isolated transaction for send and receive | Persistence Db
```
Comment (abparticular) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:19]:
```suggestion
AtomicSendsWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ReceiveOnly |  ✅| SQLP uses isolated transaction for receive | Persistence Db
```
Comment (abparticular) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:20]:
```suggestion
ReceiveOnly |  ✅| SQLT uses isolated transaction for receive | Transport DB
```

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+None |  ✅| No transactions | Persistence Db
```
Comment (abparticular) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:21]:
```suggestion
None |  ✅| No transactions | Persistence DB

<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
```

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+partial: Connection
+
+### With Outbox
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅|  Not supported | N/A
+AtomicSendsWithReceive |  ✅| Not supported | N/A
+ReceiveOnly |  ✅| Connection sharing via SQLT storage context | Persistence Db
```
Comment (abparticular) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:35]:
```suggestion
ReceiveOnly |  ✅| Connection sharing via SQLT storage context | Persistence DB
```

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md (hunk 1 above)
Comment (jpalac) [reply persistence/sql/sqlserver-combining-persistence-wth-transport.md]: @abparticular it seems to work ok - I tried it locally first then pushed the change.

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+## Connection behavior
+
+When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is used or not. This influences where the saga data is stored.
+
+### Without Outbox
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
+AtomicSendsWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```
Comment (abparticular) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:19]:
```suggestion
SendsAtomicWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```

**Review by abparticular — APPROVED** [review]

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+### Without Outbox
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
+SendsAtomicWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
+ReceiveOnly |  ✅| SQLT uses isolated transaction for receive | Transport DB
+None |  ✅| No transactions | Persistence DB
+
+<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
```
Comment (johnsimons) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:23]: Would this only work on Windows OSs?

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+## Connection behavior
+
+When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is used or not. This influences where the saga data is stored.
+
+### Without Outbox
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
```
Comment (johnsimons) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:18]: What is `SQLT`, according to google https://www.google.com/search?q=sqlt
I guess my point is that the acronym, is known for something else, I think we need to change it

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+SendsAtomicWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
+ReceiveOnly |  ✅| SQLT uses isolated transaction for receive | Transport DB
+None |  ✅| No transactions | Persistence DB
+
+<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
+
+#### Explicitly opting out of connection sharing when not using the Outbox
+
+When an endpoint uses SQL Persistence combined with the SQL Server Transport without the [Outbox](/nservicebus/outbox/), the persistence uses the connection and transaction context established by the transport when accessing saga data. This behavior ensures *exactly-once* message processing behavior as the state change of the saga is committed atomically while consuming of the message that triggered it.
```
Comment (johnsimons) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:27]:
```suggestion
When an endpoint uses SQL Persistence combined with the SQL Server Transport with [Outbox](/nservicebus/outbox/) disabled, the persistence uses the connection and transaction context established by the transport when accessing saga data. This behavior ensures *exactly-once* message processing behavior as the state change of the saga is committed atomically while consuming the message that triggered it.
```

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+partial: Connection
+
+### With Outbox
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅|  Not supported | N/A
```
Comment (johnsimons) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:35]: I don't see the purpose of ✅ in the table?

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+component: SqlPersistence
+reviewed: 2024-06-10
+related:
+  - persistence/sql/dialect-mysql
+  - transports/sql
+---
+
+## Connection behavior
+
+When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is used or not. This influences where the saga data is stored.
```
Comment (johnsimons) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:12]:
```suggestion
When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is enabled or disabled. This influences where the saga data is stored.
```

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md (hunk 7 above)
Comment (johnsimons) [reply persistence/sql/sqlserver-combining-persistence-wth-transport.md:23]: and if so, should we point it out

File: transports/sql/transactions_native_sqltransport_[6,).partial.md [hunk 12]
```diff
@@ -6,3 +6,6 @@ There are two available options within native transaction level:
 > This transaction is not shared outside of the message receiver. That means there is a possibility of persistent side-effects when processing fails, i.e. [ghost messages](/nservicebus/concepts/glossary.md#ghost-message) might occur.
 
  * **SendsAtomicWithReceive** - This mode is similar to the `ReceiveOnly`, but transaction is shared with sending operations. That means the message receive operation and any send or publish operations are committed atomically.
+
+> [!WARNING]
+> If using [SQL Server transport](/transports/sql/) and [SQL Persistence with the Microsoft SQL Server dialect](/persistence/sql/dialect-mssql.md) with different connection strings, without [outbox](/nservicebus/outbox/) enabled, in this transaction mode [the persistence uses the connection and transaction context established by the transport when accessing saga data](/persistence/sql/sqlserver-combining-persistence-wth-transport.md).
```
Comment (johnsimons) [inline transports/sql/transactions_native_sqltransport_[6,).partial.md:11]: This does not read well, needs improvement.

Comment (johnsimons) [reply transports/sql/transactions_native_sqltransport_[6,).partial.md:11]: Maybe first step is to start the sentence with "When using ...."

File: transports/transactions.md [hunk 13]
```diff
@@ -78,6 +78,10 @@ Some transports support enlisting outgoing operations in the current receive tra
 Use the following code to use this mode:
 
 snippet: TransportTransactionAtomicSendsWithReceive
+
+> [!WARNING]
+> If using [SQL Server transport](/transports/sql/) and [SQL persistence with the Microsoft SQL Server dialect](/persistence/sql/dialect-mssql.md) with different connection strings, without [outbox](/nservicebus/outbox/) enabled, in this transaction mode [the persistence uses the connection and transaction context established by the transport when accessing saga data](/persistence/sql/sqlserver-combining-persistence-wth-transport.md).
```
Comment (johnsimons) [inline transports/transactions.md:83]: Same as before

**Review by johnsimons — CHANGES_REQUESTED** [review]

File: transports/sql/transactions_native_sqltransport_[6,).partial.md (hunk 12 above)
Comment (jpalac) [reply transports/sql/transactions_native_sqltransport_[6,).partial.md]: I have reworded this to make it simpler.

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md (hunk 10 above)
Comment (jpalac) [reply persistence/sql/sqlserver-combining-persistence-wth-transport.md]: I have removed them.

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md (hunk 8 above)
Comment (jpalac) [reply persistence/sql/sqlserver-combining-persistence-wth-transport.md]: I have changed it to SQL Transport

File: persistence/sql/sqlserver-combining-persistence-wth-transport.md [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+### Without Outbox
+
+SQL Transport<br/>TransactionMode | Connection sharing | Saga location
+:-:|:-:|:-:
+TransactionScope | SQL Transport transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
+SendsAtomicWithReceive | SQL Transport uses isolated transaction for send and receive | Transport DB
+ReceiveOnly | SQL Transport uses isolated transaction for receive | Transport DB
+None | No transactions | Persistence DB
+
+<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
```
Comment (johnsimons) [inline persistence/sql/sqlserver-combining-persistence-wth-transport.md:23]:
```suggestion
<sup>1</sup> - Requires .NET Framework, or .NET 8 [`ImplicitDistributedTransactions`](https://learn.microsoft.com/en-us/dotnet/api/system.transactions.transactionmanager.implicitdistributedtransactions?view=net-8.0) set to `true`.
```

**Review by johnsimons — APPROVED** [review]

**Pull request merged**